test/topology_custom/test_raft_recovery_entry_loss.py

```diff
@@ -0,0 +1,152 @@
+from test.topology.util import check_system_topology_and_cdc_generations_v3_consistency, \
+        check_token_ring_and_group0_consistency, delete_discovery_state_and_group0_id, delete_raft_group_data, \
+        reconnect_driver, wait_for_cdc_generations_publishing
+from test.topology_custom.test_group0_schema_versioning import get_group0_schema_version, get_local_schema_version
+
+
+@pytest.mark.asyncio
+async def test_raft_recovery_entry_lose(manager: ManagerClient):
+    """
+    Test that the Raft-based recovery procedure works correctly if some committed group 0 entry has been permanently
+    lost (it has been committed only by dead nodes).
+
+    1. Start a cluster with 5 nodes
+    2. Create a scenario where nodes have the following group 0 states:
+    - node 1: v1,
+    - node 2: v2,
+    - nodes 3-5: v3.
+    3. Kill nodes 3-5 causing a permanent group 0 majority loss. We also permanently lose v3. After recovering majority,
+    nodes 1-2 should have v2, which is the only safe option.
+    4. Run the recovery procedure to recreate group 0 with nodes 1, 2 as the only members. We make sure node 2
+    becomes the leader of new group 0 since it has a newer group 0 state. When node 1 joins new group 0, it receives
+    a group 0 snapshot with v2 from node 2 and applies it. If node 1 became the leader, node 1 would end up with v1 and
+    node 2 would end up with v2. The snapshot transfer from node 1 to node 2 would not do anything due to mutations from
+    v2 having later timestamps.
+    5. Check that node 1 has moved its group 0 state to v2.
+    6. Remove nodes 3-5 from topology using the standard removenode procedure.
+    7. Add a new node (a sanity check verifying that the cluster is functioning properly).
+    """
+    logging.info('Adding initial servers')
+    servers = await manager.servers_add(5)
+    live_servers = servers[:2]
+    dead_servers = servers[2:]
+    logging.info(f'Servers to survive majority loss: {live_servers}, servers to be killed: {dead_servers}')
+
+    cql, _ = await manager.get_ready_cql(servers)
+
+    first_group0_id = (await cql.run_async(
+            "SELECT value FROM system.scylla_local WHERE key = 'raft_group0_id'"))[0].value
+
+    logging.info(f'Stopping {live_servers[0].server_id} to keep its group 0 state in v1')
+    await manager.server_stop(live_servers[0].server_id)
+
+    logging.info('Creating keyspace ks1, moving the group 0 state to v2')
+    await cql.run_async(
+            "CREATE KEYSPACE ks1 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
+
+    logging.info(f'Stopping {live_servers[1].server_id} to keep its group 0 state in v2')
+    await manager.server_stop(live_servers[1].server_id)
+
+    logging.info('Creating keyspace ks2, moving the group 0 state to v3')
+    await cql.run_async(
+            "CREATE KEYSPACE ks2 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
```

Comment:
Do we have a guarantee that the nodes being shut down will see the v2/v3 before that? Perhaps we should ask them to perform a read barrier.

Reply:
We probably don't have the guarantee. I'll add read barriers.

Reply:
done

Reply:
done